File: server/src/game/HeroPower.ts
import type { Game } from './Game.js';
import type { Player } from './Player.js';
import { Minion } from './Minion.js';
import { makeUniqueCardId } from './Deck.js';
import { MAX_BOARD_SIZE } from './constants.js';
import { SILVER_HAND_RECRUIT_TOKEN as recruit } from './tokenCards.js';
import { BoardFullError, InvalidTargetError } from './errors.js';

export interface IHeroPower {
  validate(player: Player): void;
  execute(game: Game, player: Player): void;
}
class DamagePower implements IHeroPower {
  constructor(private readonly amount: number) {}
  validate(): void {}
  execute(game: Game, player: Player): void {
    player.recordDamage(game.getOpponent().heroState.takeDamage(this.amount));
  }
}
class RecruitPower implements IHeroPower {
  validate(player: Player): void {
    if (player.boardCount >= MAX_BOARD_SIZE) throw new BoardFullError();
  }
  execute(_game: Game, player: Player): void {
    player.summonMinion(new Minion(makeUniqueCardId(), recruit.id, recruit.name,
      recruit.attack, recruit.health, player.id, false, recruit.imagePath));
  }
}
class HealPower implements IHeroPower {
  validate(): void {}
  execute(_game: Game, player: Player): void { player.heroState.heal(2); }
}
class DrawPower implements IHeroPower {
  validate(): void {}
  execute(_game: Game, player: Player): void {
    player.heroState.takeDamage(2);
    if (player.deckSize > 0) player.addToHand(player.drawCard());
  }
}
const powers: Record<string, IHeroPower> = {
  MAGE: new DamagePower(1), HUNTER: new DamagePower(2),
  PALADIN: new RecruitPower(), PRIEST: new HealPower(), WARLOCK: new DrawPower(),
};
export function heroPowerFor(heroClass: string): IHeroPower {
  const power = powers[heroClass];
  if (!power) throw new InvalidTargetError('Hero chưa được hỗ trợ.');
  return power;
}

File: server-java/src/main/java/vn/coincard/server/game/HeroPower.java
package vn.coincard.server.game;

import java.util.List;
import java.util.UUID;

/** Hero-power strategy registry. */
public interface HeroPower {
  HeroClass supportedClass();
  void validate(Player player);
  void execute(Game game, Player player);

  static List<HeroPower> defaults() {
    return List.of(
        new DamagePower(HeroClass.MAGE, 1),
        new DamagePower(HeroClass.HUNTER, 2),
        new RecruitPower(),
        new HealPower(),
        new DrawPower());
  }

  class DamagePower implements HeroPower {
    private final HeroClass heroClass;
    private final int amount;
    DamagePower(HeroClass heroClass, int amount) {
      this.heroClass = heroClass;
      this.amount = amount;
    }
    @Override public HeroClass supportedClass() { return heroClass; }
    @Override public void validate(Player player) {}
    @Override public void execute(Game game, Player player) {
      player.recordDamage(game.getOpponent().heroState().takeDamage(amount));
    }
  }

  class RecruitPower implements HeroPower {
    @Override public HeroClass supportedClass() { return HeroClass.PALADIN; }
    @Override public void validate(Player player) {
      if (player.boardCount() >= Constants.MAX_BOARD_SIZE) throw new GameException.BoardFull();
    }
    @Override public void execute(Game game, Player player) {
      CardTypes.CardDefinition recruit = TokenCards.SILVER_HAND_RECRUIT_TOKEN;
      player.summonMinion(new Minion(UUID.randomUUID().toString(),
          recruit.id(), recruit.name(), recruit.attack(), recruit.health(),
          player.id(), false, recruit.imagePath()));
    }
  }

  class HealPower implements HeroPower {
    @Override public HeroClass supportedClass() { return HeroClass.PRIEST; }
    @Override public void validate(Player player) {}
    @Override public void execute(Game game, Player player) { player.heroState().heal(2); }
  }

  class DrawPower implements HeroPower {
    @Override public HeroClass supportedClass() { return HeroClass.WARLOCK; }
    @Override public void validate(Player player) {}
    @Override public void execute(Game game, Player player) {
      player.heroState().takeDamage(2);
      player.drawForTurn();
    }
  }
}
